src/mask.c

```diff
@@ -2215,13 +2220,31 @@ void mask_init(struct db_main *db, char *unprocessed_mask)
 
 	/* Handle command-line (or john.conf) masks given in UTF-8 */
 	if (options.input_enc == UTF_8 && options.internal_cp != UTF_8) {
-		if (valid_utf8((UTF8*)mask) > 1)
+		if (valid_utf8((UTF8*)mask) > 1) {
+			int u8_length = strlen8((UTF8*)mask);
+
 			utf8_to_cp_r(mask, mask, strlen(mask));
-		for (i = 0; i < MAX_NUM_CUST_PLHDR; i++)
-			if (valid_utf8((UTF8*)options.custom_mask[i]) > 1)
-				utf8_to_cp_r(options.custom_mask[i],
-				             options.custom_mask[i],
-				             strlen(options.custom_mask[i]));
+			if (strlen(mask) != u8_length) {
+				if (john_main_process)
+					fprintf(stderr, "Error: The selected internal codepage can't hold all characters of mask\n");
+				error();
+			}
+		}
+
+		for (i = 0; i < MAX_NUM_CUST_PLHDR; i++) {
+			if (valid_utf8((UTF8*)options.custom_mask[i]) > 1) {
+				int u8_length = strlen8((UTF8*)options.custom_mask[i]);
+				char *tmp_buf;
+
+				tmp_buf = mem_alloc(strlen(options.custom_mask[i]));
```

Comment:
Shouldn't it be `strlen(...) + 1`?

Reply:
The conversion is guaranteed at this point to be one or more bytes shorter, but we might want to add a comment.

Reply:
Yes, at least a comment would help.

Also, I'd prefer to have and consistently use APIs accepting the destination buffer size and refusing to overflow the buffer, except possibly in performance-critical places.

Reply:
You mean the strcpy? It's simply **known** to be safe at that point, but I can change it anyway - it's not performance critical at all.

Reply:
Fixed